android-components/components/feature/search/src/main/java/mozilla/components/feature/search/telemetry/BaseSearchTelemetry.kt

```diff
@@ -21,20 +25,47 @@ import mozilla.components.support.base.facts.collect
 import mozilla.components.support.base.log.logger.Logger
 import mozilla.components.support.ktx.kotlinx.coroutines.flow.filterChanged
 import org.json.JSONObject
+import java.io.File
+
+internal const val COLLECTION_NAME = "search-telemetry-v2"
 
 /**
  * Main configuration and functionality for tracking ads / web searches with specific providers.
  */
-abstract class BaseSearchTelemetry {
+abstract class BaseSearchTelemetry(private val file: File) {
+    @VisibleForTesting
+    internal var providerList: List<SearchProviderModel>? = emptyList()
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    internal fun initializeProviderList(): Job {
+        val fetcher = SerpTelemetryFetcher(file, COLLECTION_NAME)
+
+        return CoroutineScope(Dispatchers.IO).launch {
+            providerList = fetcher.fetchSearchProviders()
+        }
+    }
 
     /**
-     * Install the web extensions that this functionality is based on and start listening for updates.
+     * Finds provider among list of providers that matches regex in url.
      */
-    abstract fun install(engine: Engine, store: BrowserStore)
+    @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
+    fun getProviderForUrl(url: String): SearchProviderModel? {
+        return providerList?.find { provider ->
+            provider.searchPageRegexp.pattern.toRegex().containsMatchIn(url)
+        }
+    }
 
     @VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)
-    internal fun getProviderForUrl(url: String): SearchProviderModel? =
-        providerList.find { provider -> provider.regexp.containsMatchIn(url) }
+    internal fun waitUntilProviderListInitialized() {
+        if (providerList.isNullOrEmpty()) {
+            initializeProviderList()
+        }
+    }
+
+    /**
+     * Install the web extensions that this functionality is based on and start listening for updates.
+     */
+    abstract fun install(engine: Engine, store: BrowserStore)
```

Comment:
I still think we should change this `install` function to be `suspend`ing, like `abstract suspend fun`. This will communicate more clearly to callers that the function expects to do async operations. Creating a coroutine scope inside `installWebExtension` to do the work hides some of that semantic value, IMO.

Reply:
That's a good point!